app/controllers/courses_controller.rb

```diff
@@ -17,138 +18,66 @@ def index
+   else
       @pending_grading = {}
     end
     # elsif @registration.staff?
     #   @pending_grading = []
     #   @assignments = []
-    @allocated_grading =
-      GraderAllocation
-      .where(grade_id: current_user.id)
-      .where(course: @course)
-      .where(grading_completed: nil)
-      .group_by(&:assignment_id)
-    @completed_grading =
-      GraderAllocation
-      .where(grade_id: current_user.id)
-      .where(course: @course)
-      .where.not(grading_completed: nil)
-      .joins(:submission)
-      .where("submissions.score": nil)
-      .group_by(&:assignment_id)
+    @allocated_grading = @course.grading_assigned_for(current_user)
+    @completed_grading = @course.grading_done_for(current_user)
     @assignments = Assignment
                    .where(id: @pending_grading.keys + @allocated_grading.keys + @completed_grading.keys)
                    .map{|a| [a.id, a]}.to_h
 
     if current_user_prof_for?(@course)
-      @abnormals = {}
-      people = @course.users_with_drop_info.where("registrations.dropped_date IS ?", nil).to_a
-      @assns = @course.assignments.to_a.sort_by(&:due_date)
-      all_subs = Assignment.submissions_for(people, @assns).group_by(&:assignment_id)
-      used_subs = UsedSub.where(assignment_id: @assns.map(&:id)).group_by(&:assignment_id)
-      @assns.each do |a|
-        next unless all_subs[a.id]
-        a_subs = all_subs[a.id].group_by(&:for_user)
-        used = used_subs[a.id]
-        people.each do |p|
-          subs = a_subs[p.id]
-          if (subs and subs.count > 0) and (used.nil? or used.find{|us| us.user_id == p.id}.nil?)
-            @abnormals[a] = [] unless @abnormals[a]
-            @abnormals[a].push p
-          end
-        end
-      end
-      @unpublished =
-        Submission
-        .joins("INNER JOIN used_subs ON submissions.id = used_subs.submission_id")
-        .where("used_subs.assignment_id": @assns.map(&:id))
-        .joins("INNER JOIN grades ON grades.submission_id = submissions.id")
-        .where.not("grades.score": nil)
-        .where("grades.available": false)
-        .joins("INNER JOIN users ON used_subs.user_id = users.id")
-        .order("users.name")
-        .select("DISTINCT submissions.*", "users.name AS user_name")
+      @abnormals = @course.abnormal_subs
+      @assns = @course.assignments_sorted
+      @unpublished = @course.unpublished_grades
     end
   end
 
   def new
-    new_with_errors nil
+    @course = Course.new
+    @course.lateness_config = LatenessConfig.default
+    @course.sections = [Section.new(instructor: current_user), Section.new]
+    render :new, layout: 'application'
   end
 
   def edit
-    prep_sections
+    @course.sections << Section.new
   end
 
   def facebook
     unless current_user_site_admin? || current_user_staff_for?(@course)
-      redirect_back fallback_location: root_path, alert: "Must be an admin or professor."
+      redirect_back root_path, alert: "Must be an admin or professor."
```

Comment:
Why evict the user all the way out to the root?  Just dump them back where they were.  This should be fallback_location, not root_path.